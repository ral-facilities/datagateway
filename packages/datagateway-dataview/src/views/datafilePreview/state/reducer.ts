import DATAFILE_PREVIEWER_DEFAULT from '../defaults';
import {
  DecrementDatafilePreviewerZoomLevelType,
  IncrementDatafilePreviewerZoomLevelType,
  ResetDatafilePreviewerZoomLevelType,
  ToggleDatafilePreviewerDetailsPanePayload,
  ToggleDatafilePreviewerDetailsPaneType,
} from './actions';
import { DGDataViewState } from '../../../state/app.types';

/**
 * Defines the shape of the redux state for {@link DatafilePreviewer}
 */
interface DatafilePreviewerState {
  /**
   * Whether the details pane should be shown or not.
   */
  isDetailsPaneShown: boolean;

  /**
   * The current zoom level of the previewer, in percentage. For example,
   * 100% maps to 100, 110% maps to 110, 80% maps to 80, etc.
   */
  zoomLevel: number;
}

function toggleDatafilePreviewerDetailsPane(
  state: DGDataViewState,
  payload: ToggleDatafilePreviewerDetailsPanePayload
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      isDetailsPaneShown: payload.shouldShow,
    },
  };
}

function incDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      zoomLevel:
        state.datafilePreviewer.zoomLevel +
        DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep,
    },
  };
}

function decDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      // zoom level should not go lower than zoomLevelStep
      // otherwise will result in zoom level <= 0.
      zoomLevel: Math.max(
        DATAFILE_PREVIEWER_DEFAULT.minZoomLevel,
        state.datafilePreviewer.zoomLevel -
          DATAFILE_PREVIEWER_DEFAULT.zoomLevelStep
      ),
    },
  };
}

function resetDatafilePreviewerZoomLevel(
  state: DGDataViewState
): DGDataViewState {
  return {
    ...state,
    datafilePreviewer: {
      ...state.datafilePreviewer,
      zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
    },
  };
}

const datafilePreviewerInitialState: DatafilePreviewerState = {
  isDetailsPaneShown: DATAFILE_PREVIEWER_DEFAULT.showDetailsPane,
  zoomLevel: DATAFILE_PREVIEWER_DEFAULT.zoomLevel,
};

const datafilePreviewerReducer = {
  [ToggleDatafilePreviewerDetailsPaneType]: toggleDatafilePreviewerDetailsPane,
  [IncrementDatafilePreviewerZoomLevelType]: incDatafilePreviewerZoomLevel,
  [DecrementDatafilePreviewerZoomLevelType]: decDatafilePreviewerZoomLevel,
  [ResetDatafilePreviewerZoomLevelType]: resetDatafilePreviewerZoomLevel,
};

export {
  datafilePreviewerInitialState,
  datafilePreviewerReducer,
  toggleDatafilePreviewerDetailsPane,
  incDatafilePreviewerZoomLevel,
  decDatafilePreviewerZoomLevel,
  resetDatafilePreviewerZoomLevel,
};
export type { DatafilePreviewerState };
