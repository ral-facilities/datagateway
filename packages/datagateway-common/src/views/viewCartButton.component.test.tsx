import React from 'react';
import configureStore from 'redux-mock-store';
import { initialState as dGCommonInitialState } from '../state/reducers/dgcommon.reducer';
import { StateType } from '../state/app.types';
import { Provider } from 'react-redux';
import thunk from 'redux-thunk';
import { MemoryRouter } from 'react-router-dom';
import { mount, ReactWrapper, shallow } from 'enzyme';
import { QueryClientProvider, QueryClient } from 'react-query';
import ViewCartButton, { CartProps } from './viewCartButton.component';
import { Badge } from '@mui/material';

describe('Generic cart button', () => {
  const mockStore = configureStore([thunk]);
  let state: StateType;
  let props: CartProps;

  const navigateToDownload = jest.fn();

  const createWrapper = (props: CartProps): ReactWrapper => {
    const store = mockStore(state);
    return mount(
      <Provider store={store}>
        <MemoryRouter initialEntries={[{ key: 'testKey', pathname: '/' }]}>
          <QueryClientProvider client={new QueryClient()}>
            <ViewCartButton {...props} />
          </QueryClientProvider>
        </MemoryRouter>
      </Provider>
    );
  };

  beforeEach(() => {
    props = {
      cartItems: [],
      navigateToDownload: navigateToDownload,
    };

    state = JSON.parse(
      JSON.stringify({
        dgdataview: {}, //Dont need to fill, since not part of the test
        dgcommon: {
          ...dGCommonInitialState,
          urls: {
            ...dGCommonInitialState.urls,
            idsUrl: 'https://www.example.com/ids',
          },
        },
      })
    );
  });

  afterEach(() => {
    jest.clearAllMocks();
    navigateToDownload.mockClear();
  });

  it('renders correctly', () => {
    const wrapper = shallow(<ViewCartButton {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('calls the navigate to download plugin when the cart clicked', () => {
    const wrapper = createWrapper(props);

    wrapper.find('[aria-label="app.cart_arialabel"]').last().simulate('click');

    expect(navigateToDownload).toHaveBeenCalledTimes(1);
  });

  it('has cartItems', () => {
    props = {
      cartItems: [
        {
          entityId: 1,
          entityType: 'investigation',
          id: 1,
          name: 'test',
          parentEntities: [],
        },
        {
          entityId: 2,
          entityType: 'investigation',
          id: 2,
          name: 'tes2',
          parentEntities: [],
        },
      ],
      navigateToDownload: navigateToDownload,
    };
    const wrapper = createWrapper(props);

    expect(wrapper.find(Badge).props().badgeContent).toEqual(2);
  });
});
