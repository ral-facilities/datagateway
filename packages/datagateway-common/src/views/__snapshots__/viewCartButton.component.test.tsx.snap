// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generic cart button renders correctly 1`] = `
<ForwardRef(IconButton)
  aria-label="app.cart_arialabel"
  className="tour-dataview-cart-icon"
  onClick={[MockFunction]}
  sx={
    Object {
      "margin": "auto",
    }
  }
>
  <ForwardRef(Badge)
    badgeContent={null}
    color="primary"
  >
    <Memo(ForwardRef(ShoppingCartIcon)) />
  </ForwardRef(Badge)>
</ForwardRef(IconButton)>
`;
